### PR TITLE
Do not remove calls to functions without a body during inlining

### PR DESCRIPTION
**Pull request by danpoe**
When using ``--inline`` or ``--function-inline`` in goto-instrument, calls to functions without a body are now kept (partial inlining already keeps them).

Additionally there's now a separate analysis pass ``--remove-calls-nobody`` which removes calls to functions without a body and replaces them by a nondet assignment to the return variable and evaluations of the call arguments.

### CODE REVIEW

**Comment by tautschnig** [comment]
Throughout this PR: shouldn't it be `no_body` (and `no-body`) instead of `nobody`?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 1]
```diff
@@ -683,6 +686,10 @@ int goto_instrument_parse_optionst::doit()
       status() << "Performing full inlining" << eom;
       goto_inline(goto_functions, ns, ui_message_handler);
 
+      status() << "Removing calls to functions without a body" << eom;
+      remove_calls_nobodyt rcn;
+      rcn.remove_calls_nobody(goto_functions);
```
Comment (thk123) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Probably more "cprovery" to use `operator()` so this would become:

```C++
remove_calls_nobodyt remove_calls_nobody;
remove_calls_nobody(goto_functions);
```

File: regression/goto-instrument/remove-calls-nobody2/main.c [hunk 2]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (thk123) [inline regression/goto-instrument/remove-calls-nobody2/main.c]: Nit: blank line

File: regression/goto-instrument/remove-calls-nobody1/main.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (thk123) [inline regression/goto-instrument/remove-calls-nobody1/main.c]: Nit: blank line

File: regression/goto-instrument/inline_17/main.c [hunk 4]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (thk123) [inline regression/goto-instrument/inline_17/main.c]: Nit: blank line

File: regression/goto-instrument/inline_16/main.c [hunk 5]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (thk123) [inline regression/goto-instrument/inline_16/main.c]: Nit: blank line

File: regression/goto-instrument/inline_17/main.c [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+void func1()
+{
+  func2();
+}
+
+int main()
+{
+  func1();
+}
+
```
Comment (thk123) [inline regression/goto-instrument/inline_17/main.c]: Nit: blank line

File: regression/goto-instrument/inline_16/main.c [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+void func1();
+
+int main()
+{
+  int ret;
+
+  func1();
+  ret = func2();
+}
+
```
Comment (thk123) [inline regression/goto-instrument/inline_16/main.c]: Nit: blank line

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 8]
```diff
@@ -1075,27 +1082,21 @@ void goto_instrument_parse_optionst::instrument_goto_program()
...
     status() << "Partial inlining" << eom;
-    goto_partial_inline(goto_functions, ns, ui_message_handler, true);
+    goto_partial_inline(goto_functions, ns, ui_message_handler, 0, true);
 
     goto_functions.update();
     goto_functions.compute_loop_numbers();
   }
 
-  // now do full inlining, if requested
-  if(cmdline.isset("inline"))
```
Comment (thk123) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Am I reading this right? There were two identical blocks for dealing with `cmdline.isset("inline")`?

File: src/goto-instrument/goto_instrument_parse_options.h [hunk 9]
```diff
@@ -76,7 +76,8 @@ Author: Daniel Kroening, kroening@kroening.com
   "(horn)(skip-loops):(apply-code-contracts)(model-argc-argv):" \
   "(show-threaded)(list-calls-args)(print-path-lengths)" \
   "(undefined-function-is-assume-false)" \
-  "(remove-function-body):"
+  "(remove-function-body):" \
+  "(remove-calls-nobody)"
```
Comment (thk123) [inline src/goto-instrument/goto_instrument_parse_options.h]: This flag is missing documentation, also if possible could you move the defines for it into `remove_calls_nobody.h` as described here: https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD#L114

File: src/goto-programs/goto_inline_class.cpp [hunk 10]
```diff
@@ -450,7 +450,7 @@ void goto_inlinet::insert_function_body(
   dest.destructive_insert(target, tmp);
 }
 
-void goto_inlinet::insert_function_nobody(
+void goto_inlinet::handle_function_nobody(
```
Comment (thk123) [inline src/goto-programs/goto_inline_class.cpp]: Now would be a good time to add the doxygen header to this function as it isn't clear what handling a function without a body entails.

File: src/goto-programs/remove_calls_nobody.cpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Remove calls to functions without a body
+
+#include "remove_calls_nobody.h"
+
+void remove_calls_nobodyt::remove_call_nobody(
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: Could you add Doxygen comments for all the functions?

File: src/goto-programs/remove_calls_nobody.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  const exprt::operandst &arguments)
+{
+  assert(target->is_function_call());
+  assert(!dest.empty());
+
+  goto_programt tmp;
+
+  // evaluate function arguments -- they might have
+  // pointer dereferencing or the like
+  forall_expr(it, arguments)
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: I think this can be done using a range based for which is now preferred:

```C++
for(const exprt &argument : arguments)
```

File: src/goto-programs/remove_calls_nobody.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  {
+    goto_programt::targett t=tmp.add_instruction();
+    t->make_other(code_expressiont(*it));
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: This is probably me not understanding something - but what does making an other expression do? Don't we need some kind of nondet modification if it is a pointer?

File: src/goto-programs/remove_calls_nobody.cpp [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    if(may_have_body)
+    {
+      const goto_functiont &goto_function=f_it->second;
+      may_have_body=goto_function.body_available();
+    }
+
+    if(!may_have_body)
+    {
+      remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: Is it OK to modify `goto_program` while we have an interator pointing into it?

File: src/goto-programs/remove_calls_nobody.cpp (hunk 11 above)
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: Instead of all methods being called `remove_call_nobody`, perhaps this could be called `generate_nondet_method_stub`

File: src/goto-programs/remove_calls_nobody.cpp [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+  // kill call
+  target->type=LOCATION;
+  target->code.clear();
+  target++;
+
+  dest.destructive_insert(target, tmp);
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: Instead of all being called `remove_calls_nobody` maybe this could be called `remove_no_body_calls_from_function`

File: src/goto-programs/remove_calls_nobody.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    else // called function has a body
+    {
+      it++;
+    }
+  }
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(goto_functionst &goto_functions)
+{
+  Forall_goto_functions(f_it, goto_functions)
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: Range based for:

```C++
for(auto &function_entry : goto_functions)
```

File: src/goto-programs/remove_calls_nobody.cpp [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(
+  goto_programt &goto_program,
+  const goto_functionst &goto_functions)
+{
+  for(goto_programt::targett it=goto_program.instructions.begin();
+      it!=goto_program.instructions.end();) // no it++
+  {
+    if(!it->is_function_call())
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: I'd be tempted to break the logic of this loop up into two parts:

```C++
bool is_opaque_function_call(const goto_programt::target instruction, const goto_functions &goto_functions)
{
  if(!instruction->is_function_call()) 
    return false;
...
if(!goto_function.body_available())
  return true;
}
```

And then this function becomes:

```C++
for(goto_programt::targett it=goto_program.instructions.begin();
    it!=goto_program.instructions.end();) // no it++
{
  if(is_opaque_function_call(it, goto_function)
  {
    remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
  }
  else
  {
    it++;
  }
}
```

This way it is easy to see what the loop is doing and when the iterator gets advanced

File: src/goto-programs/remove_calls_nobody.cpp [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+      remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
+    }
+    else // called function has a body
+    {
+      it++;
+    }
+  }
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(goto_functionst &goto_functions)
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.cpp]: As suggested earlier, this would become `operator()(goto_functionst &goto_functions)`

File: src/goto-programs/remove_calls_nobody.h [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    goto_programt &dest,
+    goto_programt::targett target,
+    const exprt &lhs,
+    const exprt::operandst &arguments);
+
+  void remove_calls_nobody(
+    goto_programt &goto_program,
+    const goto_functionst &goto_functions);
+
+  void remove_calls_nobody(goto_functionst &goto_functions);
```
Comment (thk123) [inline src/goto-programs/remove_calls_nobody.h]: I think this method is the only one that should be public as the others are utilities that only make sense to be called in the specific way that this one calls them.

File: regression/goto-instrument/inline_17/test.desc [hunk 20]
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--function-inline func1
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+func2\(\)
+--
+^warning: ignoring
```
Comment (thk123) [inline regression/goto-instrument/inline_17/test.desc]: Should this test check `func1` isn't called?

File: regression/goto-instrument/inline_09/test.desc [hunk 21]
```diff
@@ -1,6 +1,6 @@
 CORE
 main.c
---inline
+--inline --remove-calls-nobody
```
Comment (thk123) [inline regression/goto-instrument/inline_09/test.desc]: Could we add to this test a requirement that there is no call to `f`

File: regression/goto-instrument/inline_17/test.desc (hunk 20 above)
Comment (NlightNFotis) [reply regression/goto-instrument/inline_17/test.desc]: I agree with this. `func2\(\)` appears to me like it would be in the file even if it `func1` isn't actually inlined.

File: src/goto-programs/remove_calls_nobody.cpp (hunk 14 above)
Comment (NlightNFotis) [reply src/goto-programs/remove_calls_nobody.cpp]: This is also something that I want to know.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 8 above)
Comment (danpoe) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Yes

File: src/goto-programs/remove_calls_nobody.cpp (hunk 13 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_nobody.cpp]: An other expression just amounts to evaluating the expression. For example, for a call ``func(a[i])`` we insert an evaluation of ``a[i]`` so we can still detect out of bounds errors on the array access. The intention here is not to soundly model everything the function might do with say a pointer argument but just to "simulate" the program evaluating the arguments to the call before making the call.

File: src/goto-programs/remove_calls_nobody.cpp (hunk 14 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_nobody.cpp]: Yes, that's fine. Since it's a ``std::list`` of instructions all the iterators to elements in the list remain valid when changing the list.

File: src/goto-programs/remove_calls_nobody.cpp (hunk 16 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_nobody.cpp]: This would need to be ``for(auto &function_entry : goto_functions.function_map)`` which I'm not sure would be preferable.

File: regression/goto-instrument/inline_17/test.desc (hunk 20 above)
Comment (danpoe) [reply regression/goto-instrument/inline_17/test.desc]: I think it's fine. ``--function-inline func1`` inlines all the calls that ``func1`` makes but the call to ``func1`` itself remains. I've also added a test now that ``func1`` is called.

File: regression/goto-instrument/inline_09/test.desc (hunk 21 above)
Comment (danpoe) [reply regression/goto-instrument/inline_09/test.desc]: Good point.

File: src/goto-programs/remove_calls_nobody.cpp (hunk 17 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_nobody.cpp]: Good suggestion. It made the code a lot clearer.

**Comment by danpoe** [comment]
I think I've addressed all the comments. I didn't add new tests tough as I think the current ones are sufficient, since we're not aiming to add nondet assignments to things a pointer might point to but simply to evaluate the arguments.

File: src/goto-programs/remove_calls_nobody.cpp (hunk 16 above)
Comment (thk123) [reply src/goto-programs/remove_calls_nobody.cpp]: I would prefer it since I think it is more idiomatic, someone new to the code doesn't need to check what Forall_goto_functions does, but I don't feel strongly about it - as you prefer.

Comment (tautschnig) [reply src/goto-programs/remove_calls_nobody.cpp]: While I do agree, I am wondering whether there should be a concerted attempt to replace all occurrences of `Forall_goto_functions` in a consistent way instead of risking to introduce a multitude of variations expressing the same? Thoughts?

Comment (thk123) [reply src/goto-programs/remove_calls_nobody.cpp]: I agree in theory; in practise I think big sweeping changes tend to takes ages/never get merged/break peoples PRs etc. Small, incremental improvements in code people are actually modifying nicely prioritises code people are likely to touch and actually gets done.

File: src/goto-programs/goto_inline.cpp [hunk 22]
```diff
@@ -48,40 +48,40 @@ void goto_inline(
 
   typedef goto_functionst::goto_functiont goto_functiont;
 
-    // find entry point
```
Comment (peterschrammel) [inline src/goto-programs/goto_inline.cpp]: The style clean-ups should probably go into a separate commit.

**Comment by danpoe** [comment]
Ok, I've squashed the last commit. I've also removed the function ``handle_function_no_body()`` as it no longer warrants a separate function.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 23]
```diff
@@ -696,6 +700,10 @@ int goto_instrument_parse_optionst::doit()
       status() << "Performing full inlining" << eom;
       goto_inline(goto_functions, ns, ui_message_handler);
 
+      status() << "Removing calls to functions without a body" << eom;
+      remove_calls_no_bodyt remove_calls_no_body;
+      remove_calls_no_body(goto_functions);
+
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Why is this specifically done here - is this the only place calling `goto_inline`?

File: src/goto-programs/remove_calls_no_body.cpp [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+/// the return value.
+/// \param goto_functions: goto functions to operate on
+void remove_calls_no_bodyt::operator()(
+  goto_functionst &goto_functions)
+{
+  Forall_goto_functions(f_it, goto_functions)
+  {
+    goto_functiont &goto_function=f_it->second;
+
+    if(goto_function.body_available())
```
Comment (tautschnig) [inline src/goto-programs/remove_calls_no_body.cpp]: I don't think you even need this check - an empty body will simply make the loop terminate straight away.

File: src/goto-programs/remove_calls_no_body.h [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifndef CPROVER_GOTO_PROGRAMS_REMOVE_CALLS_NO_BODY_H
+#define CPROVER_GOTO_PROGRAMS_REMOVE_CALLS_NO_BODY_H
+
+#include "goto_functions.h"
+
+class remove_calls_no_bodyt
+{
+protected:
+  typedef goto_functionst::goto_functiont goto_functiont;
```
Comment (tautschnig) [inline src/goto-programs/remove_calls_no_body.h]: I am wondering whether this typedef is of much value - it seems to be used twice, and one of those you could even eliminate given my earlier comment.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 23 above)
Comment (danpoe) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Yes, this is the only analysis that specifically calls ``goto_inline`` before.

File: src/goto-programs/remove_calls_no_body.cpp (hunk 24 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_no_body.cpp]: Right.

File: src/goto-programs/remove_calls_no_body.h (hunk 25 above)
Comment (danpoe) [reply src/goto-programs/remove_calls_no_body.h]: Ok, I'm removing it.

Comment (tautschnig) [reply src/goto-programs/remove_calls_no_body.h]: Happy to approve once the changes have been pushed.

Comment (danpoe) [reply src/goto-programs/remove_calls_no_body.h]: Ok, I've pushed the changes.

**Comment by kroening** [comment]
Needs rebase, but can then go!

**Comment by danpoe** [comment]
@martin-cs As far as I remember we produced the Toyota releases by merging together a bunch of branches, this one being one of them. Or has the process changed meanwhile? Which branch and which patches are you referring to?

File: src/goto-programs/remove_calls_no_body.h (hunk 25 above)
Comment (tautschnig) [reply src/goto-programs/remove_calls_no_body.h]: Is that line a left-over to be removed?

File: src/goto-programs/remove_calls_no_body.cpp (hunk 24 above)
Comment (tautschnig) [reply src/goto-programs/remove_calls_no_body.cpp]: Was there an intention to make a change?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 26]
```diff
@@ -990,7 +992,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   // now do full inlining, if requested
   if(cmdline.isset("inline"))
   {
-    do_indirect_call_and_rtti_removal();
+    do_indirect_call_and_rtti_removal(/*force=*/true);
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: It might be better change the API to use an enum to make clear what the value means?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 27]
```diff
@@ -1109,29 +1111,15 @@ void goto_instrument_parse_optionst::instrument_goto_program()
...
+    status() << "Partial inlining" << eom;
+    goto_partial_inline(goto_functions, ns, ui_message_handler, 0, true);
 
-    status() << "Performing full inlining" << eom;
-    goto_inline(goto_model, get_message_handler(), true);
+    goto_functions.update();
+    goto_functions.compute_loop_numbers();
   }
 
+
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Nit pick: blank line seems unnecessary?

File: src/goto-programs/goto_inline_class.h [hunk 28]
```diff
@@ -6,7 +6,6 @@ Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
 
-
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.h]: Nit pick: change seems not too important?

**Comment by martin-cs** [comment]
On Thu, 2018-02-15 at 09:05 -0800, danpoe wrote:
> @martin-cs As far as I remember we produced the Toyota releases by merging together a bunch of branches, this one being one of them. Or has the process changed meanwhile? Which branch and which patches are you referring to?
@danpoe A few things have changed.  Let's get this merged and then have a look.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 26 above)
Comment (danpoe) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: I think having boolean flags for options should be fine.

Comment (tautschnig) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: It surely is, but the fact that you felt compelled to add a comment suggests that an enum (with just two values!) could be helpful for people reading the code.

Comment (danpoe) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: I agree that it would be more readable. I'm slightly concerned about consistency with the rest of the code base though as having two-value enums for boolean flags isn't done commonly there. I believe I've added the comment explaining the argument here just randomly (I think it's also the case for lots of other function calls with boolean literals as arguments that it's not clear what the flag does unless one inspects the declaration/definition). I'll remove the comment from the call.

Comment (tautschnig) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: I know that #1547 introduced one of those, and there was some PR (not sure whether that one or another) that got me very confused initially. Indeed it's not yet common practice, but likely we should look into adopting it. Because I'm in the mood: #1854.
